feat: add AWS SigV4 proxy integration

nickvelloff opened the pull request:
## Summary
  - introduce the AWS SigV4 provider (docs, spec entry, logo, generated templates, and connect UI copy) so customers can create proxy integrations against AWS services
  - add the SigV4 auth flow on the backend (new /auth/aws-sigv4 controller, session/reconnect handling, validation, and reconnect defaults) plus expose the new auth type through the shared types & SDKs
  - add the signing/proxy helpers (shared proxy service + unit tests) and hook them into the connection service so signed calls pick up the correct base URL and ARN context

  ## Testing
  - npm run test packages/shared/lib/services/proxy/aws-sigv4.unit.test.ts
  - npm run test packages/server/test/integration/connect/postReconnect.integration.test.ts
  - npm run test packages/server/test/integration/providerConfigKey/patchIntegration.integration.test.ts
<!-- Summary by @propel-code-bot -->

---

In addition to introducing end-to-end `AWS_SIGV4` support, this PR also strengthens configuration security and lifecycle handling by separating sensitive SigV4 credentials from public integration config, applying encrypted secret storage, and ensuring redaction/normalization in integration responses. It further expands the experience with dedicated AWS SigV4 settings behavior in the web app (including built-in vs custom STS configuration) while preserving reconnect/default flows and legacy compatibility expectations.

---
*This summary was automatically generated by @propel-code-bot*

nick-braintrust committed:
feat: add aws sigv4 integration

nick-braintrust committed:
Validate AWS SigV4 custom config and pass reconnect defaults

nick-braintrust committed:
Fix AWS SigV4 base URL derivation

TBonnin commented:
A few questions:
- who is running/providing the STS helper service?
- can you share the cloudformation template?

arsalazarjr committed:
fix: add missing AWS SigV4 error types and type fixes

agusayerza commented on packages/shared/lib/auth/aws-sigv4.ts:
```diff
@@ -0,0 +1,147 @@
+}
+
+export function getAwsSigV4Settings(config: ProviderConfig): Result<AwsSigV4IntegrationSettings, NangoError> {
+    const rawSettings = config.custom?.[AWS_SIGV4_CUSTOM_KEY];
```
STS endpoint auth secrets (API key, basic auth password) are stored inside an unstructured JSON blob in the custom field rather than in dedicated secret fields, right? Those secrets shouldn't be buried in a nested JSON structure inside `custom` as they can't be individually rotated, audited, or access-controlled separately from the rest of the integration config. Plus, it would be super simple to accidentally expose them.

agusayerza commented on packages/server/lib/controllers/auth/postAwsSigV4.ts:
```diff
@@ -0,0 +1,365 @@
+            return;
+        }
+
+        const externalId = (connectionConfig['external_id'] as string) || uuidv4();
```
We shouldn't accept the `external_id` provided by the client without validation. ([The External ID is a critical AWS confused deputy protection](https://docs.aws.amazon.com/IAM/latest/UserGuide/confused-deputy.html)). If an attacker passes a known `external_id` that matches one already trusted in a victim's role trust policy, they could assume that role. I think for new connections Nango should always generate the `external_id` server-side. For reconnections, it should only reuse the `external_id` already stored in the existing connection's `connection_config`, never accept a new one from the client.

agusayerza commented on packages/shared/lib/services/proxy/utils.ts:
```diff
@@ -208,7 +212,17 @@ export function buildProxyURL({ config, connection }: { config: ApplicationConst
+        const awsService = awsCreds.service || (connection.connection_config?.['service'] as string | undefined);
+
+        if (awsRegion && awsService) {
+            apiBase = `https://${awsService}.${awsRegion}.amazonaws.com`;
```
There’s no integration-level base URL for AWS SigV4. The provider has no `proxy.base_url`, so `buildProxyURL` hardcodes `https://${service}.${region}.amazonaws.com`. That breaks for services with non-standard endpoint patterns (e.g., S3 uses bucket/region or global endpoints and has path-style vs virtual-hosted style) aren’t supported unless callers pass `baseUrlOverride` on *every proxy call*. We should add an integration-level override (or partition-aware resolution) so this is reusable/configurable across AWS environments.

agusayerza commented on packages/webapp/src/pages/Integrations/providerConfigKey/Settings/components/General.tsx:
```diff
@@ -310,6 +678,271 @@ export const SettingsGeneral: React.FC<{
                 </div>
             )}
+
+            {template.auth_mode === 'AWS_SIGV4' && (
```
Let's extract this to a component so we don't add ~300 lines of AWS-specific UI directly in `General.tsx`, which is the shared settings component for all auth types

agusayerza commented on packages/webapp/src/pages/Integrations/providerConfigKey/Settings/components/General.tsx:
```diff
@@ -310,6 +678,271 @@ export const SettingsGeneral: React.FC<{
+                        <div className="flex flex-col gap-4">
+                            <div className="grid grid-cols-2 gap-4">
+                                <div className="flex flex-col gap-1">
+                                    <label className="text-xs text-white font-semibold">AWS Service</label>
```
Hardcoded color values violate project conventions. Instead of `text-white` use `text-text-primary`

agusayerza commented on packages/webapp/src/pages/Integrations/providerConfigKey/Settings/components/General.tsx:
```diff
@@ -310,6 +678,271 @@ export const SettingsGeneral: React.FC<{
+                                <div className="flex flex-col gap-4">
+                                    {awsSigV4Config?.templates && awsSigV4Config.templates.length > 0 ? (
+                                        awsSigV4Config.templates.map((template, index) => (
+                                            <div key={`template-${index}`} className="border border-subtle rounded-md p-4 flex flex-col gap-3 bg-[#0f1117]">
```
Hardcoded color `bg-[#0f1117]`

agusayerza commented on docs/spec.yaml:
```diff
@@ -91,6 +91,7 @@ paths:
                     $ref: '#/components/responses/NotFound'
 
         post:
+            operationId: postAuthAwsSigV4
```
I agree with the bot here, we should remove this line. This has only been added to `/auth/aws-sigv4/{providerConfigKey}` right?

agusayerza reviewed (changes requested):
The STS endpoint design makes this non-reusable. We should support direct AWS STS `AssumeRole` calls as a built-in option. Nango should be able to call AWS STS `AssumeRole` directly using the integration owner's own AWS credentials stored in the encrypted credential fields. This would remove the need to deploy a bespoke microservice for most of use cases and make the provider reusable. The custom STS endpoint can remain as an advanced/power-user option for users who need custom validation or multi-tenant routing logic between Nango and AWS.)

arsalazarjr committed:
address PR feedback for AWS SigV4 integration

arsalazarjr committed:
Fixed errors

arsalazarjr committed:
Updated per PR feedback

arsalazarjr committed:
Move STS auth secrets to dedicated integration_secrets column

arsalazarjr committed:
Refactored AwaSigV4Settings from Shared Settings to components

arsalazarjr commented on packages/server/lib/controllers/auth/postAwsSigV4.ts:
```diff
@@ -0,0 +1,365 @@
+    }
+});
+
+async function verifyAwsCredentials({
```
Yes, we already do. Lines 387-405 parse the GetCallerIdentity response, extract the Arn, and compare the role name portion against the expected role_arn. It's a warning-only log today (not a hard failure) since the assumed-role ARN format differs from the IAM role ARN (sts::123:assumed-role/MyRole/session vs iam::123:role/MyRole).

arsalazarjr committed:
Fix PR review feedback: operationId placement, optional region, proxy base_url, SigV4 payload hashing

arsalazarjr committed:
Added built-in AWS STS mode

arsalazarjr committed:
Fixed bugs from merge

arsalazarjr committed:
Fixed clipboard issue for external id

arsalazarjr committed:
cleaned up docs code sample

arsalazarjr committed:
Honor client-provided external_id when creating new connections

arsalazarjr committed:
Fixing timestamp bug

arsalazarjr committed:
clear sts auth when no auth block provided

arsalazarjr committed:
Fix external id copy to clipboard

arsalazarjr committed:
Fix post-rebase build errors

- Add missing preconfigured variable in Go.tsx (lost during rebase)
- Add getIntegrationConfig prop to ProxyRequest in postAwsSigV4.ts
  (new required prop added by upstream)
- Install @aws-sdk/client-sqs for packages/jobs (upstream dependency)

agusayerza commented on packages/server/lib/controllers/v1/integrations/providerConfigKey/patchIntegration.ts:
```diff
@@ -74,6 +77,121 @@ export const patchIntegration = asyncWrapper<PatchIntegration>(async (req, res)
+                // Update integration_secrets based on mode
+                let parsedConfig: Record<string, any> = {};
+                try {
+                    parsedConfig = JSON.parse(value);
```
Why are we parsing the value again?
We have parsed
1. To validate its JSON
2. To extract the secrets
3. To get the `stsMode`

agusayerza commented:
Hey @arsalazarjr , thanks for addressing the feedback, the config scoping explanation makes sense (integration-level owner credentials vs. connection-level customer credentials), the rebase looks clean, and the custom STS reference in #integration-templates/443 answers the testing question well.

I will be faster to review the follow up, please let me know when done.

## CloudFormation templates:
This is out of scope for Nango's current capabilities and should be handled on the integrator's side. Let strip it. The IAM setup is a prerequisite the integrator controls, not something Nango should guide through its UI. Keeping it in means we're taking on product/onboarding logic that belongs to the customer's app. Please remove the CloudFormation template feature from the integration (the template catalog, the "Deploy in AWS Console" buttons, the `instructions` block in Connect UI). The core flow: collect Role ARN + ExternalId, call STS, store credentials, is self-contained without it.

---

Other than that, there are small nits raised by Codex and 1 by me and we are done.

arsalazarjr committed:
Merge branch 'master' into feat/aws-sig4

arsalazarjr commented:
> Hey @arsalazarjr , thanks for addressing the feedback, the config scoping explanation makes sense (integration-level owner credentials vs. connection-level customer credentials), the rebase looks clean, and the custom STS reference in #integration-templates/443 answers the testing question well.
> 
> I will be faster to review the follow up, please let me know when done.
> 
> ## CloudFormation templates:
> This is out of scope for Nango's current capabilities and should be handled on the integrator's side. Let strip it. The IAM setup is a prerequisite the integrator controls, not something Nango should guide through its UI. Keeping it in means we're taking on product/onboarding logic that belongs to the customer's app. Please remove the CloudFormation template feature from the integration (the template catalog, the "Deploy in AWS Console" buttons, the `instructions` block in Connect UI). The core flow: collect Role ARN + ExternalId, call STS, store credentials, is self-contained without it.
> 
> Other than that, there are small nits raised by Codex and 1 by me and we are done.

@agusayerza Thanks for flagging this. I want to push back because removing the CloudFormation piece now would undo work we explicitly aligned on at the start of this effort and reshape the process we've already built around.

On scope: The CloudFormation templates aren't "onboarding logic for the customer's app." They're a self-service path for the integrator's end users to provision the exact IAM role (trust policy + ExternalId) that STS AssumeRole requires. Without it, every integrator using AWS Sig4 has to write the same IAM setup documentation and guide their users through console clicks. Bundling a known-good template is the lowest-friction way to avoid misconfigured trust policies.

On "product logic": The templates are static assets + a pre-filled console URL. There's no business logic, no runtime behavior, no state to maintain. It's documentation-as-a-link, not a product feature that accrues complexity.

What I'd propose instead: If the concern is that this feels like it belongs outside the core SDK, we could keep the templates but gate them behind a provider-level flag so integrators can opt out while still allowing us to use it.

arsalazarjr committed:
Merge branch 'master' into feat/aws-sig4

arsalazarjr committed:
Merge branch 'master' into feat/aws-sig4

arsalazarjr committed:
fix(aws-sigv4): address PR review feedback

- collapse internal whitespace in canonical header values per SigV4 spec
- branch buildCanonicalUri on service to avoid S3 double-encoding of %-encoded paths
- auto-detect seconds vs milliseconds for numeric expiresAt in custom STS responses
- remove silent fallback in getStsAuthFromConfig when structured secret is malformed
- strip auth field for unknown auth types in redactAwsSigV4Secrets
- split lumped missing-field check into specific role_arn / external_id / region errors
- randomize RoleSessionName via crypto.randomBytes instead of Date.now()
- validate template URLs (HTTPS + reject private/loopback hosts) before fetch
- pass noopener,noreferrer to window.open for external link buttons
- track attemptedOperation across try/catch so reconnect failures report the correct operation
- stage integration_secrets mutations via nextSecrets in patchIntegration
- parse aws_sigv4_config once; extractSecretsFromConfig now takes a parsed object
- document FormData+S3 UNSIGNED-PAYLOAD limitation

arsalazarjr commented:
> Other than that, there are small nits raised by Codex and 1 by me and we are done.

@agusayerza 
I addressed the Codex feedback and your comment [here](https://github.com/NangoHQ/nango/pull/5041/#discussion_r3088744050) in my latest commit. Here's a summary of those changes. Please lmk if it looks good.

#### Security / correctness fixes

SigV4 canonical header whitespace — packages/shared/lib/services/proxy/aws-sigv4.ts
Canonical header values now collapse internal whitespace to a single space per the SigV4 spec, not just trim ends. Prevents SignatureDoesNotMatch for headers with repeated whitespace.

S3 canonical path double-encoding — packages/shared/lib/services/proxy/aws-sigv4.ts + test
buildCanonicalUri now branches on service: S3 gets single-encoding (pathname as-is, already encoded by the URL parser), non-S3 services keep the current double-encoding. Fixes the case where S3 object keys containing %2F were being re-encoded to %252F and returning NoSuchKey. Added a regression test that locks in the S3 signature for a %2F-containing path and asserts it differs from the non-S3 signature for the same URL.

Numeric expiresAt parsing — packages/shared/lib/auth/aws-sigv4.ts
Custom STS endpoints returning Unix seconds were interpreted as milliseconds (e.g. 1772146564 → Jan 1970), causing instant-expired credentials and refres loops. Fixed with a magnitude check: numeric values below 1e12 are treated as seconds and multiplied by 1000. Diverged from the suggested blind * 1000 because that would break endpoints already returning milliseconds.

Silent fallback in getStsAuthFromConfig — packages/shared/lib/auth/aws-sigv4.ts
Post-migration, a malformed structured secret (api_key type with empty value) now returns undefined instead of falling back to the raw custom blob.
Unknown auth types also explicitly return undefined.

Unknown-auth-type leak in redactAwsSigV4Secrets — packages/server/lib/formatters/integration.ts
Legacy custom-blob path: when parsed.stsEndpoint.auth has an unknown type or missing required secret, the auth field is now stripped entirely rather than returning the raw value.

Specific error codes for missing ARN / external ID / region — packages/shared/lib/services/connection.service.ts, packages/shared/lib/utils/error.ts
The lumped !roleArn || !externalId || !region check always returned missing_aws_sigv4_region, which was misleading for connections created before the ExternalId requirement. Split into three sequential guards; registered two new error codes: missing_aws_sigv4_role_arn, missing_aws_sigv4_external_id.

Random RoleSessionName instead of Date.now() — packages/shared/lib/auth/aws-sigv4.ts
Session name is now nango-<16 hex chars> generated via crypto.randomBytes(8) — collision-free in practice, matches the convention in packages/shared/lib/auth/signature.ts. Dropped the timestamp. Diverged from the suggested Math.random() fix because it's not cryptographically secure.

Client-side SSRF guard on template fetch — packages/webapp/src/pages/Integrations/providerConfigKey/Settings/components/AwsSigV4Settings.tsx
loadAwsSigV4TemplateFromUrl now rejects non-HTTPS URLs and literal private/loopback/link-local addresses (including 169.254.169.254 — the AWS instance metadata endpoint). Added an isPrivateOrLocalHost helper covering IPv4 RFC1918, CGNAT, loopback, IPv6 unique-local and link-local. DNS-based bypass still possible, so the comment notes the server must not trust fetched template bodies.

window.open reverse-tabnapping — packages/connect-ui/src/views/Go.tsx
Both external-link buttons (awsSigV4Instructions.url and the "Deploy in AWS Console" launchUrl) now pass 'noopener,noreferrer' features to window.open.

Creation-failure hook mislabels reconnects — packages/server/lib/controllers/auth/postAwsSigV4.ts
The catch block previously reported operation: 'creation' unconditionally. Now tracks attemptedOperation across the try/catch: starts 'unknown', becomes 'creation'/'override' after the existing-connection lookup, and is refined to storedConnection.operation after upsert.

#### Refactors (same-behavior cleanups from earlier review rounds):

Parse aws_sigv4_config value once, reuse everywhere —packages/server/lib/controllers/v1/integrations/providerConfigKey/patchIntegration.ts, packages/shared/lib/auth/aws-sigv4.ts, packages/shared/lib/auth/aws-sigv4.unit.test.ts

Previously value was parsed three times in the caller (for validation, for secret extraction, and for stsMode lookup). Changed extractSecretsFromConfig to accept an already-parsed Record<string, any> instead of a raw JSON string. Caller now parses once, handles parse-failure locally, and reuses the parsed object throughout. Deleted the unused deprecated extractStsAuthFromConfig wrapper. Updated unit tests to pass parsed objects (removed the now-moot invalid-JSON test).

Stage integration_secrets mutations inside the body.custom loop — packages/server/lib/controllers/v1/integrations/providerConfigKey/patchIntegration.ts

Previously integration.integration_secrets was mutated directly across several branches inside the loop, so any early-return after the first mutation could leave integration in a partially-applied state. Introduced a nextSecrets staging variable and a secretsTouched flag; the assignment integration.integration_secrets = nextSecrets only happens once, after the loop completes, and only if any branch actually touched secrets (preserving the "don't overwrite when no aws_sigv4 key was in the patch" behavior).

#### Documented (intentionally minimal change)

FormData + S3 UNSIGNED-PAYLOAD — packages/shared/lib/services/proxy/aws-sigv4.ts
resolveSigV4Payload returns null for FormData, which falls into the S3 UNSIGNED-PAYLOAD branch. Extended the comment in hashPayload to document that signed FormData uploads to S3 are not supported; callers needing signed multipart uploads must serialize to a Buffer before signing. Treating this as a known limitation rather than a silent edge case; full FormData signing deferred as a separate design decision.

#### Pushed back on

"Builtin-mode null-creds silent success" in patchIntegration.ts — false positive
The review claimed that when builtinCredentials is null and existingStsCredentials is also null, validation passes because simulatedSecrets "does include sts_credentials," leading to a silent bad write. simulatedSecrets only injects sts_credentials when builtinCredentials is truthy:

```
  const simulatedSecrets = builtinCredentials
      ? { ...existingSecrets, aws_sigv4: { ...existingAwsSigV4, sts_credentials: builtinCredentials } }
      : existingSecrets;
```

When builtinCredentials is null, simulatedSecrets = existingSecrets (no sts_credentials), so getBuiltinCredentialsFromConfig returns null and validation fails with missing_aws_sigv4_builtin_credentials before the secret-write logic runs. The fallthrough path the review describes is unreachable.

"Hard delete destroys live connection on update" in postAwsSigV4.ts — partially incorrect
The review said hardDelete runs on update-validation failure. It doesn't:

```
  if (storedConnection.operation === 'creation') {
      await connectionService.hardDelete(storedConnection.connection.id);
  }
```

hardDelete is gated on 'creation' only. What is true: on update-validation failure, the new (failing) credentials have already overwritten the old ones via upsertAuthConnection, with no rollback — the user is stuck with broken credentials. This is a real UX issue but present in six other auth controllers (postTba, postApiKey, postTwoStep, postBasic, postSignature, postOAuth2CC). Also, validateConnection requires the DB row (uses connection.id), so simply moving validation before the upsert isn't possible. Better addressed as a coordinated PR across all auth controllers.

agusayerza commented on packages/connect-ui/src/views/Go.tsx:
```diff
@@ -103,7 +116,16 @@ export const Go: React.FC = () => {
 
-    const preconfigured = session && integration ? session.integrations_config_defaults?.[integration.unique_key]?.connection_config || {} : {};
+    const preconfiguredCredentials = session && integration ? session.integrations_config_defaults?.[integration.unique_key]?.credentials || {} : {};
+    const preconfiguredParams = session && integration ? session.integrations_config_defaults?.[integration.unique_key]?.connection_config || {} : {};
```
On reconnect we are not populating this value and thus we show a new External id. Then, when actually creating the connection we reuse the old external id (correctly). We should probably populate `session.integrations_config_defaults.[integration.unique_key]?.connection_config` with the known `external_id`

agusayerza commented on packages/server/lib/controllers/auth/postAwsSigV4.ts:
```diff
@@ -0,0 +1,422 @@
+            connection_config: connectionConfig,
+            logCtx
+        });
+
+        const [storedConnection] = await connectionService.upsertAuthConnection({
+            connectionId,
+            providerConfigKey,
+            credentials,
+            connectionConfig,
+            metadata: {},
+            config,
+            environment
+        });
```
We are dropping `connectSession?.tags`, please pipe them. 

```suggestion
        const [storedConnection] = await connectionService.upsertAuthConnection({
            connectionId,
            providerConfigKey,
            credentials,
            connectionConfig,
            metadata: {},
            config,
            environment,
            tags: connectSession?.tags
        });
```

agusayerza commented on packages/shared/lib/services/proxy/utils.ts:
`config.headers` are the caller-provided request headers, and they’re merged here after the `AWS_SIGV4` branch has already called `signAwsSigV4Request()`. So headers like `x-amz-target` can be sent on the final request without being included in the canonical request `SignedHeaders`, which makes the SigV4 signature invalid for APIs that require them.

agusayerza commented on packages/database/lib/migrations/20260224120000_add_integration_secrets_to_configs.cjs:
```diff
@@ -0,0 +1,11 @@
+exports.up = async function (knex) {
+    return knex.schema.alterTable('_nango_configs', function (table) {
+        table.jsonb('integration_secrets').nullable();
```
This column is encrypted at rest

agusayerza commented on packages/server/lib/controllers/connect/getSession.ts:
```diff
@@ -82,19 +82,52 @@ export const getConnectSession = asyncWrapper<GetConnectSession>(async (req, res
+                    ...storedConfig
+                }
+            };
+        }
```
This will leak secrets. We should not spread the full stored `connection_config` into the browser response. Only hydrate `AWS_SIGV4`, and only via an explicit allowlist of safe fields:
- `connection_config.external_id`
- `credentials.role_arn`
- `credentials.region`
Same issue exists in the `integrations_config_defaults` mapping above: even for `AWS_SIGV4` we should not echo the raw `credentials` blob, only the known non-sensitive fields.

Let's add test coverage for this as well:
1. AWS reconnect returns only:
- `connection_config.external_id`
- `credentials.role_arn`
- `credentials.region`
2. AWS reconnect does not echo extra credential keys like `secret_access_key`, `session_token`, etc.
3. Reconnect for a provider with secret `connection_config` does not echo those stored secret fields.
4. Non-`AWS_SIGV4` `integrations_config_defaults.credentials` are not echoed at all.

agusayerza committed:
refactor(integrations): move AWS SigV4 onto integration_config + clean up

Share the generic dynamic settings/create forms instead of a bespoke editor: add visible_when
conditional fields, generic secret masking (drops redactAwsSigV4Secrets), and flat SigV4 config.

nick-braintrust committed:
feat: add AWS SigV4 proxy integration

Add a new `aws-sigv4` provider and `AWS_SIGV4` auth mode that lets Nango
sign and proxy requests to AWS APIs.

- Auth: STS assume-role (with external ID) and built-in static credentials
  modes; SigV4 request signing in the proxy layer.
- Settings UI: configure region, service, STS endpoint, role ARN and
  external ID per integration.
- Connect UI: end-user credential entry for the SigV4 flow.
- Session: pass reconnect defaults and echo only known-safe AWS_SIGV4
  fields back from getSession.
- Security: server-side SSRF guard on the custom STS endpoint URL.
- Docs: provider page, connect guide, and regenerated docs catalog.

Co-authored-by: Arnie Salazar <arnie@cloudlife.io>

agusayerza committed:
Clean up

agusayerza committed:
Remove ARN from log

agusayerza committed:
Update LLMs file
